protocol/cannon-upgrades.md

```diff
@@ -0,0 +1,134 @@
+This makes the upgrade to cannon MT or 64bit work just like any other prestate upgrade.
+
+As the game type isn't changed, users would not need to reprove their withdrawals.
+
+This is really only feasible if we maintain support for legacy cannon in the monorepo until after the upgrade (and any
+existing games are completed). The legacy code could likely shortly after chains have upgraded to the new version. At
+worst chains would have to upgrade the prestate to support the next hard fork.
+
+#### State Format
+
+Detecting the VM type is going to be very expensive with the current JSON based snapshot format as it takes a number of
+seconds to parse it and it would likely need to be parsed twice - once to find the field that defines the version and
+once to parse it as that version. This could be resolved by moving to a binary state format that encodes the specific
+version at the start of the file for easy access. In doing this we could also design the format to be able to be
+streamed to make it more efficient to parse and serialize than the JSON version. That may be required with the move to
+64bit anyway due to the increase in potential client memory size creating much larger snapshots.
+
+Support for the existing JSON format could be retained for backwards compatibility (switching version based on the
+file extension similar to how gzip is enabled by file extension).
+
+`op-challenger` would need some updates to handle the different file formats. This would also allow a nice cleanup of
+challenger to stop it importing code from cannon to parse the state. It can be made almost entirely agnostic to the
+state format by executing cannon to calculate the witness hash of a snapshot rather than parsing it itself.
+
+# Proposed Solution
+
+While it complicates development slightly, I recommend shipping cannon MT and 64-bit together under the existing game
+type. The single-threaded cannon stays as 32-bit while multi-threaded moves to 64-bit. The VM type is encoded at the
+start of a new binary state format with backwards compatibility for loading JSON states.
+
+The move to multithreaded already meant introducing a number of abstractions to share code and shows that cannon can
+support multiple VM implementations at once smoothly. By shipping both versions at once we reduce the number of versions
+we need to support and simplify testing.
+
+The extra development time required to preserve backwards compatibility is likely to be offset by not being blocked on
+updating OptimismPortal to allow changing game types without invalidating withdrawals. We've also seen multiple times
+through fault proofs that developing in a backwards compatible way pays big dividends for testing and flexibility in
+deploying. With a major project underway to upgrade chains to fault proofs we want to make it as simple as possible to
+run `op-challenger` correctly and keep its deployment simple.
+
+# Risks & Uncertainties
+
+## Cost of Backwards Compatibility
```

Comment:
I'm quite ok with trivial duplication - creating a tag is effectively the same duplication as we're still maintaining that fork, it just doesn't benefit from all our usual tooling and needs special cases for releases.

The big thing that worries me about just breaking backwards compatibility is that it means that our `develop` branch isn't really releasable - you have to do some extra stuff when deploying `op-challenger` to ensure it can play extra games which every one who runs a challenger has to then do. And we're very regularly frustrated by the fact that develop isn't always releasable because that complicates how you do a release. It affects more than just security releases of the legacy cannon, we wouldn't be testing how `op-challenger` and `op-program` work with the legacy cannon since it would exist in the monorepo anymore so we then can't release those which blocks any hard forks, or we have to build some testing framework for backwards compatibility.